service/topology_coordinator.cc

```diff
@@ -1455,9 +1455,32 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                             co_return;
                         }
                         auto sched_time = tinfo.repair_task_info.sched_time;
-                        auto primary = tmap.get_primary_replica(gid.tablet);
-                        auto dst = primary.host;
                         auto tablet = gid;
+                        auto hosts_filter = tinfo.repair_task_info.repair_hosts_filter;
+                        auto dcs_filter = tinfo.repair_task_info.repair_dcs_filter;
+                        const auto& topo = _db.get_token_metadata().get_topology();
+                        std::optional<locator::host_id> dst_opt;
+                        if (hosts_filter.empty() && dcs_filter.empty()) {
+                            auto primary = tmap.get_primary_replica(gid.tablet);
+                            dst_opt = primary.host;
+                        } else {
+                           for (auto& replica : tinfo.replicas) {
+                                if (!hosts_filter.empty() && !hosts_filter.contains(replica.host)) {
+                                    continue;
+                                }
+                                auto dc = topo.get_datacenter(replica.host);
+                                if (!dcs_filter.empty() && !dcs_filter.contains(dc)) {
+                                    continue;
+                                }
+                                dst_opt = replica.host;
+                                break;
+                           }
+                        }
+                        if (!dst_opt) {
+                            throw std::runtime_error(fmt::format("Could not find host to perform repair tablet={} replica={} hosts_filter={} dcs_filter={}",
```

Comment:
There's no guarantee that every tablet will have at least 1 replica which matches the filter. Is anything preventing this failure? 

Failing here will block the state machine until manual intervention.

Reply:
Even if we do the check before submitting the request, it is possible when the locator::tablet_transition_stage::repair runs, there is no valid replicas after the filter. 

I think what we can do is:
1) if the filter is enabled:
Skip the row level repair and move to the next stage in case there is less than 2 replicas after the filter. Ensure the repair time is not updated (we will not update with current code in this case already). 

2) if the filter is not enabled
Throw as what we do now. This is not supposed to happen. 

@Deexie

Reply:
I do not understand how it is supposed to block the state machine. It runs under `advance_in_background`, won't `tablet_state.repair` just hold the failed future that will be handled here:
https://github.com/Deexie/scylla/blob/5545289bfae0508fd72b655ddd2fef6c5dbe60a2/service/topology_coordinator.cc#L1441-L1449
?

Reply:
Anyway, currently such a repair will be retried forever, so I guess this should be replaced with `co_return`, no?

Reply:
> I do not understand how it is supposed to block the state machine. It runs under `advance_in_background`, won't `tablet_state.repair` just hold the failed future that will be handled here: https://github.com/Deexie/scylla/blob/5545289bfae0508fd72b655ddd2fef6c5dbe60a2/service/topology_coordinator.cc#L1441-L1449 ?

Right, it will just make the repair will be retried forever.

Reply:
> Anyway, currently such a repair will be retried forever, so I guess this should be replaced with `co_return`, no?

I think so, yes.

I can see the logic is duplicated here and in the repair handler. Would be nice to abstract it and share it to avoid discrepancies.

E.g: 

```
for (auto& replica : tinfo.replicas) {
   if (selected_by_filter(replica, tinfo.repair_task_info, topo)) {

```

and:

```
auto dst_opt = get_primary(replica, tinfo.repair_task_info, topo);
```

...which combines the filter.

Also, in case we have a DC filter, we lost the round-robin selection of primary replica, we will always select the first one. We should probably bring it back.

Reply:
https://github.com/scylladb/scylladb/pull/23101